### PR TITLE
An allocation constraint mechanism, that de-prioritizes nodes from getting picked for allocation if they breach certain constraints

### LINKED ISSUE #487: Constraints to de-prioritize nodes from becoming shard allocation targets
**Is your feature request related to a problem? Please describe.**

In clusters where each node has a reasonable number of shards, when a node is replaced, (or a small number of node(s) added); all shards of newly created indexes get allocated on the new empty node(s). It puts a lot of stress on the single (or few) new node, deteriorating overall cluster performance.

This happens because the index-balance factor in weight function is unable to offset shard-balance when other nodes are filled with shards. It causes the new node to always have minimum weight and get picked as the target for allocation (until the new node has approx. mean number of shards/node in the cluster).

This was originally submitted as an issue to [Elasticsearch](https://github.com/elastic/elasticsearch/issues/43350).

**Describe the solution you'd like**

We propose an allocation constraint mechanism, that de-prioritizes nodes from getting picked for allocation if they breach certain constraints. Whenever an allocation constraint is breached for a shard (or index) on a node, we add a high positive constant to the node's weight. This increased weight makes the node less preferable for allocating the shard. Unlike deciders, however, this is not a hard filter. If no other nodes are eligible to accept shards (say due to deciders like disk watermarks), the shard can still be allocated on nodes with breached constraints.

**Describe alternatives you've considered**

A work around for this is to set the total-shards-per-node index/cluster setting , which limits shards of an index on a single node.

This however, is a hard limit enforced by Allocation Deciders. If breached on all nodes, the shards go unassigned causing yellow/red clusters. Configuring this setting requires careful calculation around number of nodes, and must be revised when the cluster is scaled down.

**Additional context**

https://github.com/elastic/elasticsearch/issues/43350

### PR DESCRIPTION
**Pull request by ashwinpankaj**
Signed-off-by: Ashwin Pankaj <appankaj@amazon.com>

### Description
An allocation constraint mechanism, that de-prioritizes nodes from getting picked for allocation if they breach certain constraints

### Issues Resolved
#487
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by ashwinpankaj** [comment]
@itiyamas @shwetathareja @vigyasharma

This PR is continuation of https://github.com/opensearch-project/OpenSearch/pull/489.
I had to create a new PR after the repo went public.

**Comment by nknize** [comment]
Is this still a WIP? This is also a new feature, correct? And are you wanting to target the 1.0 release?

**Review by matznermatzner — APPROVED** [review]

**Comment by ashwinpankaj** [comment]
Hi @nknize ,

> Is this still a WIP? 
Yes - I was just waiting for final sign-offs from reviewers.

> This is also a new feature, correct?
Yes

> And are you wanting to target the 1.0 release?
Yes

**Review by itiyamas — COMMENTED** [review]
Thanks for making the changes. 
General comment on tests: We have used quite determinstic exact values in tests. I felt that there is scope for adding [randomization](http://blog.mikemccandless.com/2011/03/your-test-cases-should-sometimes-fail.html) in testing as these tests run quite frequently and some of them with every commit. 

A simple example is we need not provision exact replica count per primary. It can be just a ratio. Assertions can also be re-written to match the randomization.

File: server/src/main/java/org/opensearch/cluster/routing/allocation/AllocationConstraints.java
Comment (ashwinpankaj) [reply server/src/main/java/org/opensearch/cluster/routing/allocation/AllocationConstraints.java:65]: sure - added [here](https://github.com/opensearch-project/OpenSearch/pull/680/files#diff-38fd557126b8726fed9e9914005f0854ac9ce89d5565f2a726f3ad31df600620R49)

**Comment by itiyamas** [comment]
Looks good, Ashwin. There is more scope in hotspots and constraint tests for randomization, but I will let you decide that since the tests are pretty exhaustive.

**Comment by itiyamas** [comment]
start gradle check

**Comment by setiah** [comment]
@ashwinpankaj could you also raise a backport PR on 1.x if this needs to be backported? The 1.0 freeze is today eod just fyi. If missed, this can go in 1.1.

I would let @itiyamas approve and merge this PR to main.

**Comment by ashwinpankaj** [comment]
@setiah @itiyamas does not have merge permissions AFAIK.
Would request someone with permissions to merge it to main.

**Comment by itiyamas** [comment]
I have merge permissions, but I am waiting for one more review on this PR.
@shwetathareja Please help review this.

**Comment by shwetathareja** [comment]
> Looks good, Ashwin. There is more scope in hotspots and constraint tests for randomization, but I will let you decide that since the tests are pretty exhaustive.

yes +1 @ashwinpankaj we can create a separate issue to evaluate adding randomization in tests and proceed with merge on this one.

**Review by shwetathareja — APPROVED** [review]

File: server/src/test/java/org/opensearch/cluster/routing/allocation/IndexShardHotSpotTests.java
Comment (ViggoC) [inline server/src/test/java/org/opensearch/cluster/routing/allocation/IndexShardHotSpotTests.java:30]: Hi @ashwinpankaj, would you mind explaining why you need to do exactly the same operation twice, and compare the `moves`? Thanks!

**Pull request merged**
